### PR TITLE
parser: support built-in function pow

### PR DESCRIPTION
**Pull request by sllt**

### CODE REVIEW

File: parser/scanner.l [hunk 1]
```diff
@@ -850,6 +852,10 @@ year_month		{y}{e}{a}{r}_{m}{o}{n}{t}{h}
 {outer}			return outer
 {password}		lval.item = string(l.val)
 			return password
+{pow}			lval.item = string(l.val)
+			return pow
+{power}			lval.item = string(l.val)
```
Comment (qiuyesuifeng) [inline parser/scanner.l]: Alignment, please.

File: expression/builtin/builtin.go [hunk 2]
```diff
@@ -55,8 +55,10 @@ var Funcs = map[string]Func{
 	"coalesce": {builtinCoalesce, 1, -1, true, false},
 
 	// math functions
-	"abs":  {builtinAbs, 1, 1, true, false},
-	"rand": {builtinRand, 0, 1, true, false},
+	"abs":   {builtinAbs, 1, 1, true, false},
+	"rand":  {builtinRand, 0, 1, true, false},
+	"pow":   {builtinPow, 2, 2, true, false},
+	"power": {builtinPower, 2, 2, true, false},
```
Comment (qiuyesuifeng) [inline expression/builtin/builtin.go]: Please keep in alphabetical order.

File: expression/builtin/math_test.go [hunk 3]
```diff
@@ -43,3 +43,39 @@ func (s *testBuiltinSuite) TestRand(c *C) {
 	c.Assert(v, Less, float64(1))
 	c.Assert(v, GreaterEqual, float64(0))
 }
+
+func (s *testBuiltinSuite) TestPow(c *C) {
```
Comment (qiuyesuifeng) [inline expression/builtin/math_test.go]: Please add some error and null value test case.

**Comment by qiuyesuifeng** [comment]
@sllt Thanks very much.
LGTM

File: expression/builtin/math_test.go [hunk 4]
```diff
@@ -43,3 +43,69 @@ func (s *testBuiltinSuite) TestRand(c *C) {
...
+	t = []interface{}{nil, 3}
+	_, err = builtinPow(t, nil)
+	c.Assert(err, NotNil)
+	t = []interface{}{3, nil}
+	_, err = builtinPow(t, nil)
+	c.Assert(err, NotNil)
+
+}
+
+func (s *testBuiltinSuite) TestPower(c *C) {
```
Comment (siddontang) [inline expression/builtin/math_test.go]: duplicated test code with TestPow, you should abstract them.

File: expression/builtin/math.go [hunk 5]
```diff
@@ -63,3 +63,22 @@ func builtinRand(args []interface{}, ctx map[interface{}]interface{}) (v interfa
...
+	y, err := types.ToFloat64(args[1])
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return math.Pow(x, y), nil
+
+}
+
+func builtinPower(args []interface{}, ctx map[interface{}]interface{}) (v interface{}, err error) {
```
Comment (shenli) [inline expression/builtin/math.go]: power is an alias of pow? If so, we do not need bultinPower function. builtinPow is enough?

File: parser/scanner.l [hunk 6]
```diff
@@ -850,6 +852,10 @@ year_month		{y}{e}{a}{r}_{m}{o}{n}{t}{h}
 {outer}			return outer
 {password}		lval.item = string(l.val)
 			return password
+{pow}			lval.item = string(l.val)
```
Comment (shenli) [inline parser/scanner.l]: pow and power are not reserved keywords in MySQL. So they can be used as identifiers. You need to add them to the rule of NotKeywordToken. Then they can be used as identifiers.
https://github.com/pingcap/tidb/blob/master/parser/parser.y#L1685

File: parser/parser.y [hunk 7]
```diff
@@ -2257,6 +2259,16 @@ FunctionCallNonKeyword:
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string)), Args: $3.([]ast.ExprNode)}
 	}
+|	"POW" '(' Expression ',' Expression ')'
```
Comment (shenli) [inline parser/parser.y]: Please add some unit tests in parser/parser_test.go TestBuiltin.
https://github.com/pingcap/tidb/blob/master/parser/parser_test.go#L329

**Comment by shenli** [comment]
LGTM
Thanks! @sllt

**Pull request merged**